# ChIPSeq-Analysis
I have used Snakemake based pipeline to analyse ChIP-seq data.

I have modified scripts from https://github.com/crazyhottommy/pyflow-ChIPseq to make snakefile run on the cluster at our facility.

Useful Resources:  
snakemake:  
https://hpc-carpentry.github.io/hpc-python/11-snakemake-intro/

snakemake + ChIP-seq:  
https://snakemake.readthedocs.io/en/stable/tutorial/basics.html
https://snakemake.readthedocs.io/en/stable/tutorial/advanced.html
https://molb7621.github.io/workshop/Classes/snakemake-tutorial.html

Helpful github repositories:  
https://github.com/crazyhottommy/pyflow-ChIPseq
https://github.com/ENCODE-DCC
https://gist.github.com/davfre/8596159

ChIP-seq analysis:  
http://biocluster.ucr.edu/~rkaundal/workshops/R_feb2016/ChIPseq/ChIPseq.html
https://ngschool.eu/node/51
https://github.com/shenlab-sinai/chip-seq_preprocess
https://github.com/mforde84/ENCODE_TF_ChIP_pipeline
https://www.ebi.ac.uk/training/online/course/ebi-next-generation-sequencing-practical-course/gene-regulation/chip-seq-analysis
http://www.biologie.ens.fr/~mthomas/other/chip-seq-training/

Practical Guidelines for the Comprehensive Analysis of ChIP-seq Data  
http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003326

